beartype/typing.py

```diff
@@ -288,3 +280,197 @@
     Set = set  # type: ignore[misc]
     Tuple = tuple  # type: ignore[assignment]
     Type = type  # type: ignore[assignment]
+
+if _IS_PYTHON_AT_LEAST_3_8:
+    # TODO: I'm thinking maybe this belongs in its own file that is imported
+    # here
+
+    # Define our own caching protocol
+    from typing import Protocol as _Protocol
+    from typing import runtime_checkable as runtime_checkable
+    # TODO: This is...a wart. Obviously, there is a fragility here. We should
+    # discuss.
+    from typing import _get_protocol_attrs as _get_protocol_attrs
+
+    _T_co = TypeVar("_T_co", covariant=True)
+    _TT = TypeVar("_TT", bound="type")
+
+    if TYPE_CHECKING:
+        # Warning: Deep typing voodoo ahead. See
+        # <https://github.com/python/mypy/issues/11614>.
+        from abc import ABCMeta as _ProtocolMeta
+    else:
+        _ProtocolMeta = type(_Protocol)
+
+    # TODO: Rename this?
```

Comment:
I kept the leading `_` for now to not run afoul of the name checking in `test_api_typing.py`. I'm not quite sure I understand what's going on there or how to adapt to it, but we can discuss….